packages/config-plugins/README.md

```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).
+    ├── package.json ➡️ The `main` file will be used if `index.expo-plugin.js` doesn't exist.
+    ├── app.plugin.js ➡️ ✅ `module.exports = (config) => config` -- must export a function.
+    ╰── build/index.js ➡️ ❌ Ignored because `index.expo-plugin.js` exists. This could be used with `expo-splash-screen/build/index.js`
+```
+
+#### Project folder
+
+`'./my-config-plugin'`
+
+This is different to how Node module's work because `app.plugin.js` won't be resolved by default in a folder. You'll have to manually specify the file to use it.
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── my-config-plugin/ ➡️ Folder containing plugin code
+    ╰── index.js ➡️ ✅ By default, node resolves a folder's index.js file as the main file.
+```
+
+#### Module internals
+
+If a file inside a Node module is specified, then the module's root `app.plugin.js` resolution will be skipped. This is referred to as "reaching inside a package" and is bad form. We support this to make testing, and plugin authoring easier.
+
+- `'expo-splash-screen/build/index.js'`
+- `'expo-splash-screen/build'`
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).
+    ├── package.json ➡️ The `main` file will be used if `index.expo-plugin.js` doesn't exist.
+    ├── app.plugin.js ➡️ ❌ Ignored because the reference reaches into the package internals.
+    ╰── build/index.js ➡️ ✅ `module.exports = (config) => config`
+```
+
+#### Functions
+
+You can also just pass in a config plugin.
+
+```js
+const withCustom = (config, props) => config;
+
+const config = {
+  plugins: [
+    [
+      withCustom,
+      {
+        /* props */
+      },
+    ],
+    // Without props
+    withCustom,
+  ],
+};
+```
+
+One caveat to using functions instead of strings is that serialization will replace the function with the function's name. This keeps manifests working as expected.
+
+```json
+{
+  "plugins": [["withCustom", {}], "withCustom"]
+}
+```
+
+### Why app.plugin.js for plugins
+
+Config resolution searches for a `app.plugin.js` first when a Node module name is provided.
+This is because node environments are often different to iOS, Android, or web JS environments and therefore require different transpilation presets (ex: `module.exports` instead of `import/export`).
+
+Because of this reasoning, the root of a Node module is searched instead of right next to the `index.js`. Imagine you had a TypeScript Node module where the transpiled main file was located at `build/index.js`, if Expo config plugin resolution searched for `build/app.plugin.js` you'd lose the ability to transpile the file differently.
+
+### Creating static plugins
+
+- The root file can be any JS file or the root `app.plugin.js` in a Node module.
```

Comment:
not sure what the right wording here is but i think this is the first time mentioning "root file" so we might want to reword this a bit. maybe like "The entry point to your plugin can be ..."